src/ipsecondarydiscovery.cc

```diff
@@ -0,0 +1,94 @@
+#include "ipsecondarydiscovery.h"
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <chrono>
+
+#include "logging.h"
+
+std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::discover() {
+  sendRequest();
+  return waitDevices();
+}
+
+std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::waitDevices() {
+  std::vector<Uptane::SecondaryConfig> secondaries;
+  int socket_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (socket_fd == -1) {
+    perror("socket creation");
+  }
+
+  int reuse = 1;  // Neded for tests, because client and server runs on the same machine and bind the same port
+  setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof reuse);
+
+  struct sockaddr_in recv_addr {};
+  recv_addr.sin_family = AF_INET;
+  recv_addr.sin_port = htons(config_.ipdiscovery_port);
+  recv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+  if (bind(socket_fd, (struct sockaddr *)&recv_addr, sizeof recv_addr) < 0) {
+    perror("bind");
+  }
+
+  auto start_time = std::chrono::system_clock::now();
+  struct timeval tv;
+  tv.tv_sec = config_.ipdiscovery_wait_seconds;
+  tv.tv_usec = 0;
+  setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+
+  char rbuf[2000] = {};
+  int recieved = 0;
+  while ((recieved = recv(socket_fd, rbuf, sizeof(rbuf) - 1, 0)) != -1) {
+    std::string data(std::move(rbuf), recieved);
+    try {
+      int type = 0;
+      Uptane::SecondaryConfig conf;
+      asn1::Deserializer des(data);
+      des >> asn1::seq >> asn1::implicit<kAsn1Integer>(type) >> asn1::implicit<kAsn1Utf8String>(conf.ecu_serial) >>
+          asn1::implicit<kAsn1Utf8String>(conf.ecu_hardware_id) >> asn1::restseq;
```

Comment:
@OYTIS do you think it would be worth it to also transmit the secondary server tcp port, to make the discovery more fail-safe? So that we only need to change the port in one place. We can worry about it later but it's pretty straightforward.

Reply:
We can't use discovery with tcp because it requires connection to hosts. But we need to find those hosts.

Reply:
Yes, I meant transmitting the tcp port number of the server (secondary) in the udp discovery response so that it doesn't need to be hard-coded on the primary side.

Reply:
Do we need this IP now? It is not a problem to add any fields we need later.

Reply:
You're right, I think it can wait until the integration